lib/compress/index.js

```diff
@@ -3989,8 +3989,7 @@ merge(Compressor.prototype, {
     AST_Scope.DEFMETHOD("var_names", function varNames() {
         var var_names = this._var_names;
         if (!var_names) {
-            const isParentScopeAvailable = this.parent_scope && !this.parent_scope instanceof AST_Toplevel;
-            this._var_names = var_names = isParentScopeAvailable ? varNames.call(this.parent_scope) : Object.create(null);
+            this._var_names = var_names = this.parent_scope instanceof AST_Defun ? varNames.call(this.parent_scope) : Object.create(null);
```

Comment:
This would break with block scopes! You can't assume that the parent scope is going to be a function, it might be a `.block_scope` property of any AST_Block.

Reply:
Will check and fix it later

Reply:
I tried to put function into [block](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/block) and it does not break anythink for me.. 😕 

**Command:**
`terser -c passes=3`

**Input:**
```javascript
var a = {};
var condition = false;

(function(global) {
    if (condition === true) {
        global.fail = function(o) {
            var result = {};

            function inner() {
                return outer({
                    one: 1,
                    two: o.two
                });
            }

            result.inner = function() {
                return inner();
            };

            return result;
        };
    } else {
        global.fail = function(o) {
            var result = {};

            function inner() {
                return outer({
                    one: o.one,
                    two: 2
                });
            }

            result.inner = function() {
                return inner();
            };

            return result;
        };
    }

    function outer(o) {
        console.log(o);
    }
})(a);

var b = a.fail({});
b.inner();
```

**Output:**
```javascript
var a={},condition=!1;!function(global){function outer(o){console.log(o)}a.fail=!0===condition?function(o){return{inner:function(){return outer({one:1,two:o.two})}}}:function(o){return{inner:function(){return outer({one:o.one,two:2})}}}}();var b=a.fail({});b.inner();
```

Reply:
Also checked with **input:**
```javascript
var a = {};
var condition = false;

(function (global) {
    function fail(o) {
        var result = {};

        if (condition)
            var inner = function () {
                return outer({
                    one: 1,
                    two: o.two
                });
            };
        else
            var inner = function () {
                return outer({
                    one: o.one,
                    two: 2
                });
            };

        result.inner = function () {
            return inner();
        };

        return result;
    };

    function outer(o) {
        console.log(o);
    }

    global.fail = fail;
})(a);

var b = a.fail({});
b.inner();
```

**output:**
```javascript
var a={},condition=!1;!function(global){function outer(o){console.log(o)}a.fail=function(o){var result={};if(condition)var inner=function(){return outer({one:1,two:o.two})};else inner=function(){return outer({one:o.one,two:2})};return result.inner=function(){return inner()},result}}();var b=a.fail({});b.inner();
```

Reply:
If the condition is set to `false` initially, it will get removed by the `conditionals` option.

Reply:
Because Terser knows you're just doing `false == true` and evaluates it to `false` and removes the if and therefore the block.

Reply:
Disagree, as I know, it will not get removed if I use var and will get removed if I use const

Reply:
Good point. I see that the condition was not removed, just turned into a tern.

I completely forgot that when I implemented block scopes, I used an AST_Scope instance as the AST_Block#block_scope property. That's why your code works. The parent scope is a block scope, but because of how this was implemented, it also happens to be an AST_Scope.

My confusion was because AST_Scope is the base class for functions too, so I thought this would only work if the parent scope was a function.

My apologies.